draft-ietf-mls-protocol.md

```diff
@@ -1435,48 +1461,226 @@ all arrive at the following state:
 
-The requirement for this is that all participants know these values.
-If additional clear-text fields are attached to messages (like the counter), those
-fields MUST be protected by a signed message envelope.
+- Senders MUST only use the Application Secret once and monotonically
+  increment the generation of their secret.
+
+- Receivers SHOULD delete an Application Secret once it has been used to
+  derive the corresponding AEAD key and nonce and MAY keep those around
+  for some reasonable period.
+
+- Receivers MUST delete an Application Secret once it has been used to
+  successfully decrypt a message.
+
+[[ OPEN ISSUE: A stronger requirement could be a MUST delete Application Secret N
+as soon as N+1 has been derived.]]
+
+### Application AEAD Key Calculation
+
+The Application AEAD keying material is generated from the following
+input values:
+
+- The Application Secret value;
+- A purpose value indicating the specific value being generated;
+- The length of the key being generated.
+
+The traffic keying material is generated from an input traffic secret value using:
+
+~~~~
+write_nonce_[sender] =
+  HKDF-Expand-Label(Application_Secret,"nonce", "", nonce_length)
+
+write_key_[sender] =
+  HKDF-Expand-Label(Application_Secret,"key", "", key_length)
+~~~~
+
+Note, that because the identity of the participant using the keys to send data
+is included in the initial Application Secret, all successive updates to the
+Application secret will implicitly inherit this ownership.
+
+All the traffic keying material is recomputed whenever the underlying
+Application Secret changes.
+
+
+## Message Encryption and Decryption
+
+The Group participants MUST use the AEAD algorithm associated with
+the negotiated MLS ciphersuite to AEAD encrypt and decrypt their
+Application messages and sign them as follows:
+
+~~~~~
+    struct {
+        uint32 length;
+        opaque content[length];
+        opaque signature<0..2^16-1>;
+        uint8 zeros[length_of_padding];
+    } MLSPlaintext;
+
+    struct {
+        uint8  group[32];
+        uint32 epoch;
+        uint32 generation;
+        uint32 sender;
+        uint32 length;
+        opaque encrypted_content<0..2^32-1>;
+    } MLSCiphertext;
+~~~~~
+
+The Group identifier and epoch allow a device to know which Group secrets
+should be used and from which Epoch secret to start computing other secrets
+and keys. The participant identifier is used to derive the participant
+Application secret chain from the initial shared Application secret.
+The application generation field is used to determine which Application
+secret should be used from the chain to compute the correct AEAD keys
+before performing decryption.
+
+The signature field allows strong authentication of messages:
+
+~~~
+    struct {
+        uint8  group[32];
+        uint32 epoch;
+        uint32 generation;
+        uint32 sender;
+        uint32 length;
+        opaque content[length];
+    } MLSSignatureContent;
+~~~
```

Comment:
You don't say how this structure is to be used.

Reply:
I rephrased to "The signature used in the MLSPlaintext is computed over the MLSSignatureContent
which covers the metadata ..."